code/lib/core-server/src/utils/StoryIndexGenerator.test.ts

```diff
@@ -0,0 +1,1741 @@
+/* eslint-disable @typescript-eslint/no-shadow */
+/// <reference types="@types/jest" />;
+
+/**
+ * @jest-environment node
+ */
+
+import path from 'path';
+import fs from 'fs-extra';
+import { normalizeStoriesEntry } from '@storybook/core-common';
+import type { Indexer, NormalizedStoriesSpecifier, StoryIndexEntry } from '@storybook/types';
+import { loadCsf, getStorySortParameter } from '@storybook/csf-tools';
+import { toId } from '@storybook/csf';
+import { logger, once } from '@storybook/node-logger';
+
+import type { StoryIndexGeneratorOptions } from './StoryIndexGenerator';
+import { StoryIndexGenerator } from './StoryIndexGenerator';
+
+jest.mock('@storybook/csf-tools');
+jest.mock('@storybook/csf', () => {
+  const csf = jest.requireActual('@storybook/csf');
+  return {
+    ...csf,
+    toId: jest.fn(csf.toId),
+  };
+});
+
+jest.mock('@storybook/node-logger');
+
+const toIdMock = toId as jest.Mock<ReturnType<typeof toId>>;
+const loadCsfMock = loadCsf as jest.Mock<ReturnType<typeof loadCsf>>;
+const getStorySortParameterMock = getStorySortParameter as jest.Mock<
+  ReturnType<typeof getStorySortParameter>
+>;
+
+const storiesMdxIndexer: Indexer = {
+  test: /\.stories\.mdx$/,
+  index: async (fileName, opts) => {
+    let code = (await fs.readFile(fileName, 'utf-8')).toString();
+    const { compile } = await import('@storybook/mdx2-csf');
+    code = await compile(code, {});
+    const csf = loadCsf(code, { ...opts, fileName }).parse();
+
+    // eslint-disable-next-line no-underscore-dangle
+    return Object.entries(csf._stories).map(([key, story]) => ({
```

Comment:
it's unfortunate that we have to rely in the "internal" `_stories` field here, but that's the only way to get stories' keys.
I could also change `CsfFile.stories` to return the object instead of only the values, but I don't know if that would be considered a breaking change? Depends on if we consider `CsfFile` a public API or not.

Reply:
Why do we need the `key` here in this case if we already have a `story.id`? Or is that not guaranteed to exist?

Reply:
`key` is required in the new `IndexInput`. And `key` and `id` is not the same, here's an example:

```
{
	key: 'LoggedIn',
	id: 'example-header--logged-in',
	name: 'Logged In',
  title: 'Example/Header',
}
```

Reply:
You are right, `key` is just used to calculate `name`, `name` and `title` are used to calculate `id`.
So in this case the `key` is just to satisfy the typings that require a `key`.

I think we need a longer talk about `key`/`name`/`id`